FilePersistence/src/version_control/merge/ChangeGraph.cpp

```diff
@@ -448,28 +446,113 @@ void ChangeGraph::removeLabelConflictsBetweenChildren(Model::NodeIdType parentId
 {
-	Q_ASSERT(false);
-}
+	auto parentNode = tree->find(parentId);
+	Q_ASSERT(parentNode != NULL);
+	for (auto nodeId : labelMap.keys())
+	{
+		QString labelA;
+		QString labelB;
+		QString labelNone;
+		auto labels = labelMap.values(nodeId);
+		for (auto label : labels)
+		{
+			if (label.branch_.testFlag(MergeChange::BranchA))	labelA = label.label_;
+			else if (label.branch_.testFlag(MergeChange::BranchB))	labelB = label.label_;
+			else labelNone = label.label_;	//Base
+		}
 
-void ChangeGraph::updateOutgoingLabels(Model::NodeIdType /*parentId*/, IdToLabelMap /*labelMap*/)
-{
-	Q_ASSERT(false);
+		auto changeIt = changesForNode_.find(nodeId);
+		while (changeIt != changesForNode_.end() && changeIt.key() == nodeId)
+		{
+			switch (changeIt.value()->type())
+			{
+				case ChangeType::Stationary:
+					// Since label changes are already deleted
+					Q_ASSERT(changeIt.value()->isValueOrTypeChange());
+				case ChangeType::Insertion:
+				{	// Update labels according to the versions
+					if (changeIt.value()->branches().testFlag(MergeChange::BranchA))
+					{
+						changeIt.value()->newLabel_ = labelA;
+						labelA.clear();
+					}
+					else if (changeIt.value()->branches().testFlag(MergeChange::BranchB))
+					{
+						changeIt.value()->newLabel_ = labelB;
+						labelB.clear();
+					}
+				}
+				case ChangeType::Deletion:	// Update old label in case of deletion
+					changeIt.value()->oldLabel_ = labelNone;
+				case ChangeType::Move:
+				{
+					Q_ASSERT(changeIt.value()->newParentId() == parentId ||
+								changeIt.value()->oldParentId() == parentId);
+					if (changeIt.value()->newParentId() == parentId)	// Move In
+					{
+						// In case both moves are combined into single move, we split and add two moves
+						if (changeIt.value()->branches() == MergeChange::BranchA &&
+							 changeIt.value()->branches() == MergeChange::BranchB)
+						{
+							Q_ASSERT(!labelA.isNull() && !labelB.isNull());
+							// Update flags are handled in the function
+							splitMoveChangeForSecondLabel(changeIt.value(), {labelA, MergeChange::BranchA},
+																							{labelB, MergeChange::BranchB});
+							labelA.clear();
+							labelB.clear();
+							continue;
+						}
+
+						// Update labels according to the versions
+						if (changeIt.value()->branches().testFlag(MergeChange::BranchA))
+						{
+							changeIt.value()->newLabel_ = labelA;
+							labelA.clear();
+						}
+						else if (changeIt.value()->branches().testFlag(MergeChange::BranchB))
+						{
+							changeIt.value()->newLabel_ = labelB;
+							labelB.clear();
+						}
+					}
+					else	changeIt.value()->oldLabel_ = labelNone;	// Update old Label for Move Out
+
+					// Update Flags
+					if (changeIt.value()->newLabel() == changeIt.value()->oldLabel())
+						changeIt.value()->updateFlags_ = ChangeDescription::NoFlags;
+					else changeIt.value()->updateFlags_ = ChangeDescription::Label;
+				}
+				case ChangeType::Unclassified:
```

Comment:
change this to the more general `default:` in case someone introduces new kinds of changes in the future.